sync/sync_head.go

```diff
@@ -173,22 +174,74 @@ func (s *Syncer[H]) verify(ctx context.Context, newHead H) (bool, error) {
 	}
 
 	var verErr *header.VerifyError
-	if errors.As(err, &verErr) && !verErr.SoftFailure {
-		logF := log.Warnw
-		if errors.Is(err, header.ErrKnownHeader) {
-			logF = log.Debugw
-		}
-		logF("invalid network header",
-			"height_of_invalid", newHead.Height(),
-			"hash_of_invalid", newHead.Hash(),
-			"height_of_subjective", sbjHead.Height(),
-			"hash_of_subjective", sbjHead.Hash(),
-			"reason", verErr.Reason)
+	if !errors.As(err, &verErr) {
+		return false, nil
+	}
+
+	if verErr.SoftFailure {
+		err := s.verifySkipping(ctx, sbjHead.Height(), newHead)
+		return false, err
+	}
+
+	logF := log.Warnw
+	if errors.Is(err, header.ErrKnownHeader) {
+		logF = log.Debugw
 	}
+	logF("invalid network header",
+		"height_of_invalid", newHead.Height(),
+		"hash_of_invalid", newHead.Hash(),
+		"height_of_subjective", sbjHead.Height(),
+		"hash_of_subjective", sbjHead.Hash(),
+		"reason", verErr.Reason,
+	)
 
 	return verErr.SoftFailure, err
 }
 
+/*
+Subjective head is 500, network head is 1000.
+Header at height 1000 does not have sufficient validator set overlap,
+so the client downloads height 750 (which does have enough sufficient overlap),
+verifies it against 500 and advances the subjective head to 750.
+
+Client tries to apply height 1000 against 750 and if there is sufficient overlap,
+it applies 1000 as the subjective head.
+If not, it downloads the halfway point and retries the process.
+*/
+func (s *Syncer[H]) verifySkipping(ctx context.Context, subjHeight uint64, networkHeader H) error {
```

Comment:
From sync discussion:

In the 500(sbjHead) and 1000(netHead) example, the current algo fetches 750 and verifies 1000. However, the 750 is taken for granted without verification with 500. 

Instead, the algo should fetch 750, verify it, and set it as new sbjHead to initiate the sync loop fetching 500-750, while bifurcation continues until 1000 is reached.

Reply:
There is one more minor question to answer.  What happens to new incoming heads if bifurcation takes more than a block time? Do we discard them or queue for verification once the bifurcation finishes?